### PR TITLE
Adding cache into indexation

### PR DESCRIPTION
**Pull request by kaldoran**
And here comes the cache for the indexation process.

In this PR i also move the `__normalConditionTest(self)` function of the file  python/tools/dht/tests.py
cause i've planned ahead for a next test which would test all the feature with a real network ( with some node leaving and some getting back etc ).

If that gave trouble i could update my PR

### CODE REVIEW

**Comment by kaldoran** [comment]
@sim590 here i need you for a review :yum:

File: include/opendht/indexation/pht.h [hunk 1]
```diff
@@ -47,6 +47,15 @@ struct Prefix {
         return Prefix(*this, len);
     }
 
+    /**
+     * Method for getting the bit at the position pos.
+     * @param pos : Pos of the needed bit
+     * @return : bit at the position 'pos'
+     */
+    int getBit(size_t pos) const {
```
Comment (aberaud) [inline include/opendht/indexation/pht.h]: bool

File: include/opendht/indexation/pht.h [hunk 2]
```diff
@@ -166,6 +175,125 @@ class Pht {
     void insert(Key k, Value v, Dht::DoneCallbackSimple cb = {});
 
 private:
+    class Cache {
+    public:
+        /**
+         * Insert all needed node into the tree according to a prefix
+         * %param p : Prefix that we need to insert
+         */
```
Comment (sim590) [inline include/opendht/indexation/pht.h]: This is a minor observation: the format of documentation should be consistent with the rest of the opendht code, i.e. with [doxygen format][]. I think `%param` is not part of the standard format. Use `@param` instead. 

[doxygen format]: https://www.stack.nl/~dimitri/doxygen/manual/docblocks.html

File: include/opendht/indexation/pht.h [hunk 3]
```diff
@@ -166,6 +175,125 @@ class Pht {
...
+            leaves_.insert(std::pair<time_point, std::shared_ptr<Node>>(now, curr_node ));
+        }
+
+        /**
+         * Lookup into the tree to return the maximum prefix lenght in the cache tree
+         *
+         * @param p : Prefix that we are looking for
+         * @return  : The size of the longest prefix known in the cache between 0 and p.size_
+         */
+        int lookup(const Prefix& p) {
```
Comment (sim590) [inline include/opendht/indexation/pht.h]: Again, this is minor, but you may review english in the comments.

- line 212 `s/existe/exists/`
- line 218 `s/not exist/doesn't exist/`
- line 228 `s/leave/leaf/`
- line 233 `s/lenght/length/`
- line 281 `s/betwean/between/`
- ...

File: python/tools/dht/tests.py [hunk 4]
```diff
@@ -371,6 +478,9 @@ def _gottaGetThemAllPokeNodes(self, consumer, hashes, nodes=None):
                 for n in DhtFeatureTest.foreignNodes:
                     nodes.add(n)
 
+    def __normalConditionTest(self):
```
Comment (sim590) [inline python/tools/dht/tests.py]: This is odd since `DhtFeatureTest` inherits from `FeatureTest`, so it has `self._normalConditionTest`. Therefor, at line 574, the call is simply:
```python
self._normalConditionTest()
```
as before except that the name of the function changed. In short, `__normalConditionTest` is useless.

File: src/Makefile.am [hunk 5]
```diff
@@ -30,4 +30,4 @@ nobase_include_HEADERS = \
         ../include/opendht/dhtrunner.h \
         ../include/opendht/default_types.h \
         ../include/opendht/rng.h \
-        ../include/opendht/indexation/pht.h
+		../include/opendht/indexation/pht.h
```
Comment (sim590) [inline src/Makefile.am]: Tab issues. Since I know you use `vim`, you should make sure to `:set noexpandtab` before inserting a tab in an automake file.

File: src/indexation/pht.cpp [hunk 6]
```diff
@@ -9,14 +9,14 @@ const ValueType IndexEntry::TYPE = ValueType::USER_DATA;
...
-        std::shared_ptr<unsigned> max_common_prefix_len, bool all_values)
+        std::shared_ptr<unsigned> max_common_prefix_len, int start, bool all_values)
 {
     struct node_lookup_result {
         bool done {false};
         bool is_pht {false};
     };
 
-    auto mid = (*lo + *hi)/2;
+    auto mid = /*(start >= 0) ? start :*/ (*lo + *hi)/2;
```
Comment (sim590) [inline src/indexation/pht.cpp]: I'm sure this is not meant to be left commented out. ¯\(°_o)/¯

File: src/indexation/pht.cpp [hunk 7]
```diff
@@ -180,10 +188,10 @@ void Pht::insert(Key k, Value v, Dht::DoneCallbackSimple done_cb) {
                 entry.prefix = kp.content_;
                 entry.name = name_;
 
-                updateCanary(*final_prefix);
-                dht_->put(final_prefix->hash(), std::move(entry), done_cb);
+            updateCanary(*final_prefix);
+            dht_->put(final_prefix->hash(), std::move(entry), done_cb);
```
Comment (sim590) [inline src/indexation/pht.cpp]: Bad indentation.

File: tools/dhtnode.cpp [hunk 8]
```diff
@@ -96,7 +96,7 @@ void cmd_loop(std::shared_ptr<DhtRunner>& dht, std::map<std::string, dht::indexa
             break;
         } else if (op == "h" || op == "help") {
             print_help();
-            continue;
+            continue;;
```
Comment (sim590) [inline tools/dhtnode.cpp]: Does this compile? If so, tell me and revert this change anyway.

File: src/indexation/pht.cpp [hunk 9]
```diff
@@ -27,21 +27,26 @@ void Pht::lookupStep(Prefix p, std::shared_ptr<int> lo, std::shared_ptr<int> hi,
         }
         else if (is_leaf or *lo > *hi) {
             // leaf node
+            auto to_insert = p.getPrefix(mid);
+            cache_.insert(to_insert);
+
```
Comment (sim590) [inline src/indexation/pht.cpp]: Beware empty trailing white spaces. See [this][] or [my vimrc](https://github.com/sim590/dotfiles/blob/master/vimrc#L137).

[this]: http://vim.wikia.com/wiki/Remove_unwanted_spaces

File: include/opendht/indexation/pht.h (hunk 2 above)
Comment (kaldoran) [inline include/opendht/indexation/pht.h]: Yup i saw that, in fact i use %param on my own doc in C and forget to change it here :)
I'll make an update for that then

File: src/indexation/pht.cpp (hunk 6 above)
Comment (kaldoran) [inline src/indexation/pht.cpp]: This is not, this is in fact the main part of the code that's commented here. oops :x

File: tools/dhtnode.cpp (hunk 8 above)
Comment (kaldoran) [inline tools/dhtnode.cpp]: yup that's compile, that's interpreted as a : 
continue;
;
But there is still an error

**Comment by sim590** [comment]
Trying to compile with CMake, I get this error:
```
[ 88%] Linking CXX executable dhtchat
../libopendht.so: undefined reference to `dht::indexation::Pht::Cache::NODE_EXPIRE_TIME'
```
I think [this](http://stackoverflow.com/a/30002401) is related. Also, there are a couple of warnings at compilation time which you could get rid of. For e.g.:
```
/home/simon/prog/opendht/src/indexation/pht.cpp: At global scope:
/home/simon/prog/opendht/src/indexation/pht.cpp:12:62: warning: unused parameter ‘start’ [-Wunused-parameter]
         std::shared_ptr<unsigned> max_common_prefix_len, int start, bool all_values)
                                                              ^
/home/simon/prog/opendht/src/indexation/pht.cpp:151:18: warning: unused parameter ‘ok’ [-Wunused-parameter]
         [=](bool ok){
                  ^
/home/simon/prog/opendht/src/indexation/pht.cpp:175:50: warning: unused parameter ‘values’ [-Wunused-parameter]
         [=](std::vector<std::shared_ptr<Value>>& values, Prefix p) {
```
If the parameters are unused, just remove the name in the lambda/function signature.

File: include/opendht/indexation/pht.h [hunk 10]
```diff
@@ -166,6 +175,125 @@ class Pht {
...
+                root_ = curr_node;
+            }
+            
+            curr_node->last_reply = now;
+
+            /* Iterate through all bit of the Blob */
+            for ( i = 0; i < p.size_; i++ ) {
+
+                /* According to the bit define which node is the next one */
+                auto& next = ( p.getBit(i) ) ? (*curr_node).right_child : (*curr_node).left_child;
```
Comment (sim590) [inline include/opendht/indexation/pht.h]: You should use the syntax `pointer->element` instead of `(*pointer).element`.
It's simpler and more consistent with the rest of opendht code.

Comment (kaldoran) [inline include/opendht/indexation/pht.h]: Ok thanks for the review, the patch will come soon

Comment (kaldoran) [inline include/opendht/indexation/pht.h]: For the previous error, since i compile with autotool etc i'll try it with cmake and correct the error. Thanks.

File: include/opendht/indexation/pht.h [hunk 11]
```diff
@@ -166,6 +176,125 @@ class Pht {
...
+                    tmp_curr_node->parent = curr_node;
+                    next = tmp_curr_node;
+                    curr_node = std::move(tmp_curr_node);
+                }
+
+                curr_node->last_reply = now;
+            }
+
+            /* Insert the leaf (curr_node) into the multimap */
+            leaves_.insert(std::pair<time_point, std::shared_ptr<Node>>(now, curr_node ));
```
Comment (sim590) [inline include/opendht/indexation/pht.h]: Here's some c++11 tips. You can do this:
```cpp
leaves_.insert(std::pair<time_point, std::shared_ptr<Node>>(now, curr_node));
```
but, you can shrink it to:
```cpp
leaves_.insert(std::make_pair(now, curr_node));
```
furthermore, you can do:
```cpp
leaves_.emplace(now, curr_node);
```
Finally, *if I'm not mistaken*, the leaf would be copied constructed from `curr_node`. Then, you can enhance the code by making `std::multimap::emplace` use the move constructor of `std::shared_ptr`:
```cpp
leaves_.emplace(now, std::move(curr_node));
```

**Comment by kaldoran** [comment]
Here is the lastest commit with all correction said above.

**Comment by sim590** [comment]
@kaldoran: If you can't see the warnings when compiling and you're using autotools, you have to add the flag `--enable-debug` when running `./configure`. Otherwise, using CMake, you should see the warnings. The warnings I've talked you about are not related to the changes you've made except for one that was due to [code commented out by mistake](https://github.com/savoirfairelinux/opendht/blob/c4e090ced73eb79b7c16c5ea853d75f99c37ad0e/src/indexation/pht.cpp#L19). So don't mind removing the warnings unrelated to your PR. I'll get rid of them.

As for the error, yes, even if the build directory is clean, it happens. Make sure you have "tools compilation" enabled since it happens when trying to link with the library and tools do link with the library:

**CMake**
```sh
cmake -DOPENDHT_TOOLS=on
```
**autotools**
Make you run the `configure` script without the option `--disable-tools`.

**Comment by kaldoran** [comment]
@sim590 I see the warning, but i can't reproduce this error : 
> [ 88%] Linking CXX executable dhtchat
> ../libopendht.so: undefined reference to `dht::indexation::Pht::Cache::NODE_EXPIRE_TIME'

Even by adding some flags or whatever.
Even if i try with cmake or autotools.

**Comment by kaldoran** [comment]
The compilation error had been removed thanks to the lastest commit. 
Thanks for your help @sim590

File: python/tools/dht/tests.py [hunk 12]
```diff
@@ -17,6 +17,15 @@
...
 import networkx as nx
+import subprocess
+import re
+
+from matplotlib.ticker import FuncFormatter
+import math
+
+import numpy as np
+import matplotlib.pyplot as plt
+from matplotlib.ticker import FuncFormatter
```
Comment (sim590) [inline python/tools/dht/tests.py]: This diff shows duplicate imports. Probably happenned while resolving conflicts. You should revert this change (lines 20-28).

File: python/tools/dht/tests.py (hunk 4 above)
Comment (sim590) [inline python/tools/dht/tests.py]: Also, I see that you have copied the code from `_totallyNormalTest` into another function (`_normalConditionTest`) as if you wanted to rename it, but you have not deleted the [old one](https://github.com/savoirfairelinux/opendht/pull/56/files#diff-7590f949505d55358a56a80e319fe3baR597) (line 597).

Comment (sim590) [inline python/tools/dht/tests.py]: @kaldoran, the changes done on `_totallyNormalTest` don't seem to be used, are not complete and don't quite fit into your PR. May be you could revert this, only keep the changes relevant to the PR and make another PR that involves the complete changes around `_totallyNormalTest` that you wanted to make.

File: python/tools/dht/tests.py (hunk 12 above)
Comment (kaldoran) [inline python/tools/dht/tests.py]: Update done :) Thanks for the review.
And yup in fact that was a trouble will merging ( i suppose )

File: python/tools/dht/tests.py (hunk 4 above)
Comment (kaldoran) [inline python/tools/dht/tests.py]: Well, i'm going to revert the change then, the update will follow in a new PR

**Comment by kaldoran** [comment]
Here is the final version of my PR, i merge all the change into 2 single commit

**Comment by sim590** [comment]
Merged in 9269bd1.

**Comment by kaldoran** [comment]
Wel :( i didnt understand why you didnt use the merge button ?
Cause in the commit history, there is a double addition of my line ( once by you, once by me) 
https://github.com/savoirfairelinux/opendht/commits/index

**Comment by sim590** [comment]
@kaldoran: I saw some trailing white spaces in many files, so instead of asking you to remove them since I asked a lot yet, I pulled the branch, removed white spaces and merged in manually, hence the commit with double authorship.
